README.md
# gunicorn Cookbook

[![Build Status](https://travis-ci.org/chef-cookbooks/gunicorn.svg?branch=master)](http://travis-ci.org/chef-cookbooks/gunicorn) [![Cookbook Version](https://img.shields.io/cookbook/v/gunicorn.svg)](https://supermarket.chef.io/cookbooks/gunicorn)

Installs and configures the latest version of Gunicorn (via pip), aka `Green Unicorn`, a Python WSGI HTTP Server for UNIX. It's a pre-fork worker model ported from Ruby's Unicorn project. Includes an LWRP for managing Gunicorn config files. By default Gunicorn is installed system-wide but you can target a particular `virtualenv` by overriding the `node["gunicorn"]["virtualenv"]` attribute.

## Requirements

### Platforms

- Debian/Ubuntu
- RHEL/CentOS/Scientific/Amazon/Oracle

### Chef

- Chef 12.1+

### Cookbooks

- python

## Attributes

- `node["gunicorn"]["virtualenv"]` - the virtualenv you want to target Gunicorn installation into. The virtualenv will be created if it doesn't exist.

## Resource/Provider

This cookbook includes LWRPs for managing gunicorn config files.

### gunicorn_config

Creates a Gunicorn configuration file at the path specified. Meant to be deployed with a service init scheme/supervisor such as runit. Please see the `appliation::gunicorn` recipe for a complete working example. In depth information about Gunicorn's configuration values can be [found in the Gunicorn documentation](http://gunicorn.org/#docs).

#### Actions

- :create: create a Gunicorn configuration file.
- :delete: delete an existing Gunicorn configuration file.

#### Attribute Parameters

- path: name attribute. The path where the configuration file will be created
- template: template to use when rendering the configuration file. default is `gunicorn.py.erb` (part of this cookbook)
- cookbook: cookbook to look for template file in. default is this cookbook `gunicorn`
- listen: the socket to bind to. A string of the form: 'HOST', 'HOST:PORT', 'unix:PATH'. default is `0.0.0.0:8000` or listen on port 8000 on all interfaces.  Can also be an array of addresses.
- backlog: The maximum number of pending connections. default is `2048`
- preload_app: Whether application code should be loaded before the worker processes are forked. default is `false`
- worker_processes: The number of worker process for handling requests. default is `4`
- worker_class: The type of workers to use. default is `sync`
- worker_timeout: The number of seconds to wait before a worker is killed and restarted. default is `60`
- worker_keepalive: The number of seconds to wait for requests on a Keep-Alive connection. default is `2`
- worker_max_requests: The maximum number of requests a worker will process before restarting. default is `0` or restarts disabled
- worker_max_requests_jitter: The maximum jitter to add to the max_requests setting. default is `0`
- worker_threads: The number of worker threads for handling requests. default is `1`
- worker_connections: The maximum number of simultaneious clients. default is `1000`
- worker_graceful_timeout: Timeout for graceful workers restart. default is `30`
- limit_request_line: The maximum size of HTTP request line in bytes. default is `4094`
- limit_request_fields: Limit the number of HTTP headers fields in a request. default is `100`
- limit_request_field_size: Limit the allowed size of an HTTP request header field. default is `8190`
- server_hooks: A hash with whose values will be rendered as a [Gunicorn server hook](http://gunicorn.org/configure.html#server-hooks) callables (functions) named after the hash item's key name. default is `{}` or no serves hooks
- owner: The owner for the configuration file.
- group: The group owner of the configuration file (string or id).
- pid: A filename to use for the PID file. default is no pidfile
- accesslog: The access log file to write to.
- access_log_format: The access log format.
- errorlog: The error log file to write to.
- loglevel: The granularity of error log outputs.
- logger_class: The logger you want to use to log events in gunicorn.
- logconfig: The log config file to use.
- secure_scheme_headers: A hash containing headers and values that the front-end proxy uses to indicate HTTPS requests.
- forwarded_allow_ips: Front-end's IPs from which allowed to handle set secure headers. (comma separate).
- proc_name: A base to use with setproctitle for process naming.
- raw_env: An array of raw environment variables in the form of KEY=VALUE.  default is `nil`.

#### Example

```
# create a config with the default values
gunicorn_config "/etc/gunicorn/myapp.py" do
  action :create
end

# tweak some worker related values...we're web scale baby
gunicorn_config "/etc/gunicorn/myapp.py" do
  worker_processes 8
  backlog 4096
  action :create
end

# use the 'pre_fork' server hook to
# sleep for a second before forking
gunicorn_config "/etc/gunicorn/myapp.py" do
  server_hooks({:pre_fork => 'import time;time.sleep(1)'})
  action :create
end
```

## Usage

Simply include the recipe where you want Gunicorn installed.

## License & Authors

**Author:** Cookbook Engineering Team ([cookbooks@chef.io](mailto:cookbooks@chef.io))

**Copyright:** 2011-2016, Chef Software, Inc.

```
Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
```


CHANGELOG.md
# gunicorn Cookbook CHANGELOG

This file is used to list changes made in each version of the gunicorn cookbook.

## 2.1.0 (2017-02-28)

Added the following options to the configuration resource and template:

- raw_env
- worker_threads
- worker_max_requests_jitter
- worker_graceful_timeout
- limit_request_line
- limit_request_fields
- limit_request_field_size


## 2.0.0 (2016-09-16)

- Fix doc link in the readme
- Resolve rubocop and foodcritic warnings
- Testing updates
- Require Chef 12.1+

## v1.2.1 (2015-10-20)

- Added source_url and issues_url to the metadata
- Updated gitignore file
- Add Test Kitchen config
- Added chefignore file
- Added Chef standard rubocop config
- Added Travis CI testing
- Updated Gemfile with the latest testing deps
- Updated testing and contributing docs
- Added maintainers.md and maintainers.toml files
- Added travis and cookbook version badges to the readme
- Resolved Rubocop and Foodcritic warnings
- Updated Opscode -> Chef Software
- Added supported platforms to the metadata
- Updated Chefspec to 4.X format

## v1.2.0 (2014-09-18)

- [#11] Added ChefSpec matchers, Berksfile and Gemfile for testing dependencies

## v1.1.6 (2014-06-24)

- Removes smart quotes to prevent Encoding::InvalidByteSequenceError errors.

## v1.1.4 (2014-05-28)

- [COOK-4530] in gunicorn config lwrp enable options for secure headers and proc-name

## v1.1.2

### Improvement

- [COOK-3313](https://tickets.opscode.com/browse/COOK-3313)** - Add attributes for Gunicorn's log settings

### Bug

- [COOK-3019](https://tickets.opscode.com/browse/COOK-3019)** - Fix server hook name (`start_server` -> `when_ready`)

## v1.1.0

- [COOK-1188] - add LWRP for reusability
- [COOK-1207] - add missing service hooks
- [COOK-1431] - resolve foodcritic warnings
- [COOK-1486] - add missing call to super

## v1.0.0

- Current/initial release


resources/config.rb
#
# Author:: Seth Chisamore (<schisamo@opscode.com>)
# Cookbook:: gunicorn
# Resource:: config
#
# Copyright:: 2011-2016, Chef Software, Inc <legal@opscode.com>
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

actions :create, :delete
default_action :create

attribute :path, kind_of: String, name_attribute: true
attribute :template, kind_of: String, default: 'gunicorn.py.erb'
attribute :cookbook, kind_of: String, default: 'gunicorn'

attribute :listen, kind_of: [String, Array], default: '0.0.0.0:8000'
attribute :backlog, kind_of: Integer, default: 2048
attribute :preload_app, kind_of: [TrueClass, FalseClass], default: false

attribute :worker_processes, kind_of: Integer, default: 4
attribute :worker_class, kind_of: String, default: 'sync'
attribute :worker_timeout, kind_of: Integer, default: 60
attribute :worker_keepalive, kind_of: Integer, default: 2
attribute :worker_max_requests, kind_of: Integer, default: 0
attribute :worker_max_requests_jitter, kind_of: Integer, default: 0
attribute :worker_connections, kind_of: Integer, default: 1000
attribute :worker_threads, kind_of: Integer, default: 1
attribute :worker_graceful_timeout, kind_of: Integer, default: 30

attribute :limit_request_line, kind_of: Integer, default: 4094
attribute :limit_request_fields, kind_of: Integer, default: 100
attribute :limit_request_field_size, kind_of: Integer, default: 8190

attribute :accesslog, kind_of: String, default: nil
attribute :access_log_format, kind_of: String, default: nil
attribute :errorlog, kind_of: String, default: nil
attribute :loglevel, kind_of: String, default: nil
attribute :logger_class, kind_of: String, default: nil
attribute :logconfig, kind_of: String, default: nil
attribute :secure_scheme_headers, kind_of: Hash, default: nil
attribute :forwarded_allow_ips, kind_of: String, default: nil
attribute :proc_name, kind_of: String, default: nil

attribute :server_hooks, kind_of: Hash, default: nil, \
                         callbacks: {
                           'should contain a valid gunicorn server hook name' => ->(hooks) { Chef::Resource::GunicornConfig.validate_server_hook_hash_keys(hooks) },
                         }

attribute :raw_env, kind_of: Array, default: nil

attribute :owner, regex: Chef::Config[:user_valid_regex]
attribute :group, regex: Chef::Config[:group_valid_regex]
attribute :pid, kind_of: String

VALID_SERVER_HOOK_NAMES = [
  :on_starting, :on_reload, :when_ready, :pre_fork, :post_fork,
  :pre_exec, :pre_request, :post_request, :worker_exit
].freeze

private

def self.validate_server_hook_hash_keys(server_hooks)
  server_hooks.keys.reject { |key| VALID_SERVER_HOOK_NAMES.include?(key.to_sym) }.empty?
end
